File: apps/rust-server/src/routes/reactions.rs
use axum::{
    extract::{Path, State},
    http::StatusCode,
    response::{IntoResponse, Json, Response},
    routing::{delete, get, post},
    Router,
};
use serde::Deserialize;
use serde_json::json;
use uuid::Uuid;

use crate::auth::middleware::AuthUser;
use crate::AppState;

pub fn router() -> Router<AppState> {
    Router::new()
        .route(
            "/api/messages/:messageId/reactions",
            post(add_reaction).get(get_reactions),
        )
        .route(
            "/api/messages/:messageId/reactions/:emoji",
            delete(remove_reaction),
        )
}

#[derive(Deserialize)]
struct AddReactionBody {
    emoji: String,
}

async fn add_reaction(
    State(state): State<AppState>,
    user: AuthUser,
    Path(message_id): Path<Uuid>,
    Json(body): Json<AddReactionBody>,
) -> Response {
    if body.emoji.is_empty() {
        return (
            StatusCode::BAD_REQUEST,
            Json(json!({"error": "Emoji is required"})),
        )
            .into_response();
    }

    // Verify message exists and user has access
    let msg = sqlx::query_as::<_, (Uuid, Uuid)>(
        r#"SELECT m.id, m.conversation_id
           FROM messages m
           JOIN conversations c ON c.id = m.conversation_id
           WHERE m.id = $1 AND c.user_id = $2"#,
    )
    .bind(message_id)
    .bind(&user.id)
    .fetch_optional(&state.db)
    .await;

    if matches!(msg, Ok(None) | Err(_)) {
        return (
            StatusCode::NOT_FOUND,
            Json(json!({"error": "Message not found"})),
        )
            .into_response();
    }

    let reaction_id = Uuid::new_v4();
    let result = sqlx::query(
        r#"INSERT INTO message_reactions (id, message_id, user_id, emoji)
           VALUES ($1, $2, $3, $4)
           ON CONFLICT (message_id, user_id, emoji) DO NOTHING"#,
    )
    .bind(reaction_id)
    .bind(message_id)
    .bind(&user.id)
    .bind(&body.emoji)
    .execute(&state.db)
    .await;

    match result {
        Ok(_) => (StatusCode::CREATED, Json(json!({"success": true}))).into_response(),
        Err(e) => (
            StatusCode::INTERNAL_SERVER_ERROR,
            Json(json!({"error": e.to_string()})),
        )
            .into_response(),
    }
}

async fn remove_reaction(
    State(state): State<AppState>,
    user: AuthUser,
    Path((message_id, emoji)): Path<(Uuid, String)>,
) -> Response {
    let result = sqlx::query(
        "DELETE FROM message_reactions WHERE message_id = $1 AND user_id = $2 AND emoji = $3",
    )
    .bind(message_id)
    .bind(&user.id)
    .bind(&emoji)
    .execute(&state.db)
    .await;

    match result {
        Ok(r) if r.rows_affected() == 0 => (
            StatusCode::NOT_FOUND,
            Json(json!({"error": "Reaction not found"})),
        )
            .into_response(),
        Ok(_) => StatusCode::NO_CONTENT.into_response(),
        Err(e) => (
            StatusCode::INTERNAL_SERVER_ERROR,
            Json(json!({"error": e.to_string()})),
        )
            .into_response(),
    }
}

async fn get_reactions(
    State(state): State<AppState>,
    user: AuthUser,
    Path(message_id): Path<Uuid>,
) -> Response {
    // Verify message exists and user has access
    let msg = sqlx::query_as::<_, (Uuid,)>(
        r#"SELECT m.id
           FROM messages m
           JOIN conversations c ON c.id = m.conversation_id
           WHERE m.id = $1 AND c.user_id = $2"#,
    )
    .bind(message_id)
    .bind(&user.id)
    .fetch_optional(&state.db)
    .await;

    if matches!(msg, Ok(None) | Err(_)) {
        return (
            StatusCode::NOT_FOUND,
            Json(json!({"error": "Message not found"})),
        )
            .into_response();
    }

    // Get grouped reactions with count and whether current user reacted
    let reactions = sqlx::query_as::<_, (String, i64, bool)>(
        r#"SELECT
               emoji,
               COUNT(*)::bigint AS count,
               BOOL_OR(user_id = $2) AS user_reacted
           FROM message_reactions
           WHERE message_id = $1
           GROUP BY emoji
           ORDER BY MIN(created_at)"#,
    )
    .bind(message_id)
    .bind(&user.id)
    .fetch_all(&state.db)
    .await;

    match reactions {
        Ok(rows) => {
            let result: Vec<serde_json::Value> = rows
                .iter()
                .map(|(emoji, count, user_reacted)| {
                    json!({
                        "emoji": emoji,
                        "count": count,
                        "userReacted": user_reacted,
                    })
                })
                .collect();
            Json(json!(result)).into_response()
        }
        Err(e) => (
            StatusCode::INTERNAL_SERVER_ERROR,
            Json(json!({"error": e.to_string()})),
        )
            .into_response(),
    }
}

File: apps/server/src/routes/reactions.ts
import type { FastifyInstance } from "fastify";
import { db } from "../db/index.js";
import { messageReactions, messages, conversations } from "../db/schema.js";
import { eq, and, inArray } from "drizzle-orm";
import { requireAuth } from "../middleware/auth.js";
import { z } from "zod";

const addReactionSchema = z.object({
  emoji: z.string().min(1).max(32),
});

/** Fetch reactions for a list of messages (batch). Returns Map<messageId, reactions[]> */
export async function getReactionsForMessages(messageIds: string[]) {
  if (messageIds.length === 0) return new Map<string, typeof rows>();
  const rows = await db
    .select()
    .from(messageReactions)
    .where(inArray(messageReactions.messageId, messageIds));

  const byMsg = new Map<string, typeof rows>();
  for (const r of rows) {
    const existing = byMsg.get(r.messageId) ?? [];
    existing.push(r);
    byMsg.set(r.messageId, existing);
  }
  return byMsg;
}

export async function reactionRoutes(app: FastifyInstance) {
  // Add reaction to a message
  app.post<{
    Params: { messageId: string };
  }>("/api/messages/:messageId/reactions", async (request, reply) => {
    const user = await requireAuth(request, reply);
    const { messageId } = request.params;
    const body = addReactionSchema.parse(request.body);

    // Verify message exists and belongs to user's conversation
    const [msg] = await db
      .select({ conversationId: messages.conversationId })
      .from(messages)
      .where(eq(messages.id, messageId));

    if (!msg) {
      return reply.status(404).send({ error: "Message not found" });
    }

    const [conv] = await db
      .select({ userId: conversations.userId })
      .from(conversations)
      .where(eq(conversations.id, msg.conversationId));

    if (!conv || conv.userId !== user.id) {
      return reply.status(403).send({ error: "Forbidden" });
    }

    // Upsert reaction (ignore if already exists)
    try {
      const [reaction] = await db
        .insert(messageReactions)
        .values({
          messageId,
          userId: user.id,
          emoji: body.emoji,
        })
        .onConflictDoNothing()
        .returning();

      return reply.status(201).send(reaction ?? { messageId, userId: user.id, emoji: body.emoji });
    } catch {
      return reply.status(409).send({ error: "Reaction already exists" });
    }
  });

  // Remove reaction from a message
  app.delete<{
    Params: { messageId: string; emoji: string };
  }>("/api/messages/:messageId/reactions/:emoji", async (request, reply) => {
    const user = await requireAuth(request, reply);
    const { messageId, emoji } = request.params;

    const [deleted] = await db
      .delete(messageReactions)
      .where(
        and(
          eq(messageReactions.messageId, messageId),
          eq(messageReactions.userId, user.id),
          eq(messageReactions.emoji, decodeURIComponent(emoji))
        )
      )
      .returning();

    if (!deleted) {
      return reply.status(404).send({ error: "Reaction not found" });
    }

    return reply.status(204).send();
  });

  // Get reactions for a message
  app.get<{
    Params: { messageId: string };
  }>("/api/messages/:messageId/reactions", async (request, reply) => {
    const user = await requireAuth(request, reply);
    const { messageId } = request.params;

    const reactions = await db
      .select()
      .from(messageReactions)
      .where(eq(messageReactions.messageId, messageId));

    // Group by emoji
    const grouped: Record<string, { emoji: string; count: number; userReacted: boolean }> = {};
    for (const r of reactions) {
      if (!grouped[r.emoji]) {
        grouped[r.emoji] = { emoji: r.emoji, count: 0, userReacted: false };
      }
      grouped[r.emoji].count++;
      if (r.userId === user.id) grouped[r.emoji].userReacted = true;
    }

    return reply.send(Object.values(grouped));
  });
}
